### PR TITLE
Fix#15: Avoidance of zero division

### LINKED ISSUE #15: :multdiv always gives a NaN matrix and a zero matrix
I'm using julia 0.4.6 with NMF.jl 0.2.5. After entering the following code into Julia REPL

```
using NMF
V = rand(4, 2)
nnmf(V, 2, alg=:multdiv)
```

I got an NaN matrix and a zero matrix

```
NMF.Result{Float64}(4x2 Array{Float64,2}:
 NaN  NaN
 NaN  NaN
 NaN  NaN
 NaN  NaN,2x2 Array{Float64,2}:
 0.0  0.0
 0.0  0.0,1,true,NaN)
```

However, when I switched to the default method:

```
nnmf(V, 2)
```

it returned sensible results:

```
NMF.Result{Float64}(4x2 Array{Float64,2}:
 0.0495014  0.475371
 0.81839    0.0     
 0.064684   0.946153
 0.646644   0.094437,2x2 Array{Float64,2}:
 1.0193     0.0944043
 7.9901e-7  0.801651 ,8,true,5.866905652862745e-8
```

Am I not using it correctly or is the `:multdiv` method not implemented correctly?

### PR DESCRIPTION
**Pull request by tsano430**
I fixed #15 by adding a small positive constant `lambda` to the denominator of the update rules.

### CODE REVIEW

File: src/multupd.jl [hunk 1]
```diff
@@ -148,7 +148,7 @@ function update_wh!(upd::MultUpdDiv{T}, s::MultUpdDiv_State{T}, X, W::Matrix{T},
     mul!(WtQ, transpose(W), Q)
     sum!(fill!(sW, 0), W)
     @inbounds for j = 1:n, i = 1:k
-        H[i,j] *= (WtQ[i,j] / sW[i])
+        H[i,j] *= (WtQ[i,j] / (sW[i] + lambda))
```
Comment (ararslan) [inline src/multupd.jl:151]: I'm not particularly familiar with these algorithms. Is the regularization coefficient guaranteed to be small enough that it won't meaningfully change the values computed here?

**Review by ararslan — COMMENTED** [review]
Can you add a test for this?

File: src/multupd.jl (hunk 1 above)
Comment (tsano430) [reply src/multupd.jl:151]: No, it is not guaranteed. I'm sorry, I didn't realize that `lambda` is the regularization coefficient. However, it seems that `lambda` is neither the L1 regularization coefficient nor the L2 regularization coefficient according to [1], [2]. So, I'll modify the code as follows:

- Delete `lambda` which has unclear meaning, and define new coefficients `lambda_w` and `lambda_h`
- Assume that `lambda_w` and `lambda_h` are the L1 regularization coefficient for W and H, respectively
- Define a new small positive constant `delta` to avoid division by zero 
- Assume that `delta` is not user-specified

Also, I'll create a test on non-negative constraints, NaN, and the accuracy of the solutions. 

[1] A. Cichocki, S. Amari, R. Zdunek, R. Kompass, G. Hori and Z. He, "Extended SMART algorithms for non-negative matrix factorization,"  Proceedings of the 8th International Conference on Artificial Intelligence and Soft Computing, pp.548–562, 2006.
[2] A. Koso and N. Takahashi, "Derivation of multiplicative update rules for nonnegative matrix factorization with regularization terms," Proceedings of the 2017 International Symposium on Nonlinear Theory and Its Applications, pp.322-325, 2017.

Comment (tsano430) [reply src/multupd.jl]: cf. https://gist.github.com/tsano430/dc005fd59f89662d777774de84cd5d0d

File: src/multupd.jl [hunk 2]
```diff
@@ -7,40 +7,44 @@
...
+    tol::T          # change tolerance upon convergence
+    lambda_w::T     # L1 regularization coefficient for W
+    lambda_h::T     # L1 regularization coefficient for H
 
     function MultUpdate{T}(;obj::Symbol=:mse,
                             maxiter::Integer=100,
                             verbose::Bool=false,
                             tol::Real=cbrt(eps(T)),
-                            lambda::Real=sqrt(eps(T))) where T
+                            lambda_w::Real=zero(T),
```
Comment (mschauer) [inline src/multupd.jl:21]:
```suggestion
                            lambda_w=lambda,
```
would make things easier later on

File: src/multupd.jl [hunk 3]
```diff
@@ -7,40 +7,44 @@
...
+                            lambda_h::Real=zero(T)) where T
```
Comment (mschauer) [inline src/multupd.jl:22]:
```suggestion
                            lambda_h::Real=lambda) where T
```

**Comment by mschauer** [comment]
I am a bit lost here about the design. Can we just make reasonable default values for `lambda_w` and `lambda_h` to get rid of if-statements like
```
       if lambda_h > zero(T)
             H[i,j] *= (WtQ[i,j] / (sW[i] + lambda_h))
         else
             H[i,j] *= (WtQ[i,j] / (sW[i] + delta))
         end
```

**Comment by tsano430** [comment]
@mschauer Thank you for your review. That's a good suggestion. On the other hand, some users may be overwrite the default values to zero. So I think it's safer to modify the code (line 9-31 in `src/multupd.jl`) as follows: 
```julia
mutable struct MultUpdate{T}
    obj::Symbol     # objective :mse or :div
    maxiter::Int    # maximum number of iterations
    verbose::Bool   # whether to show procedural information
    tol::T          # change tolerance upon convergence
    lambda_w::T     # L1 regularization coefficient for W
    lambda_h::T     # L1 regularization coefficient for H

    function MultUpdate{T}(;obj::Symbol=:mse,
                            maxiter::Integer=100,
                            verbose::Bool=false,
                            tol::Real=cbrt(eps(T)),
                            lambda_w::Real=zero(T),
                            lambda_h::Real=zero(T)) where T

        obj == :mse || obj == :div || throw(ArgumentError("Invalid value for obj."))
        maxiter > 1 || throw(ArgumentError("maxiter must be greater than 1."))
        tol > 0 || throw(ArgumentError("tol must be positive."))
        lambda_w >= 0 || throw(ArgumentError("lambda_w must be non-negative."))
        lambda_h >= 0 || throw(ArgumentError("lambda_h must be non-negative."))
        if obj == :div
            lambda_w = max(lambda_w, sqrt(eps(T)))
            lambda_h = max(lambda_h, sqrt(eps(T)))
        end
        new{T}(obj, maxiter, verbose, tol, lambda_w, lambda_h)
    end
end
```

As a result, the extra if-statements can be deleted. Could I commit the change?

**Comment by mschauer** [comment]
Yes!

**Comment by tsano430** [comment]
@mschauer I committed the change, but some checks were not successful. I'm not familiar with `coverage/coveralls`, so I don't know how to remedy the problem.

**Comment by mschauer** [comment]
> 17 of 17 new or added lines in 1 file covered.

All is well.

**Review by mschauer — APPROVED** [review]

**Comment by tsano430** [comment]
@mschauer Thank you for your comment and approval!

File: src/multupd.jl [hunk 4]
```diff
@@ -7,40 +7,48 @@
...
+    verbose::Bool   # whether to show procedural information
+    tol::T          # change tolerance upon convergence
+    lambda_w::T     # L1 regularization coefficient for W
+    lambda_h::T     # L1 regularization coefficient for H
 
     function MultUpdate{T}(;obj::Symbol=:mse,
                             maxiter::Integer=100,
                             verbose::Bool=false,
                             tol::Real=cbrt(eps(T)),
-                            lambda::Real=sqrt(eps(T))) where T
```
Comment (ararslan) [inline src/multupd.jl:20]: The removal of this keyword argument in the constructor constitutes a breaking change. We should probably at least put in a deprecation for it by including `lambda=nothing` as a keyword argument and emitting a warning if a non-`nothing` value is provided (and set `lambda_w` and `lambda_h` appropriately to accommodate the user's `lambda` value).

Comment (tsano430) [reply src/multupd.jl:20]: @ararslan Thank you for your comment. I modified the code.

File: src/multupd.jl [hunk 5]
```diff
@@ -7,40 +7,54 @@
...
 
         obj == :mse || obj == :div || throw(ArgumentError("Invalid value for obj."))
         maxiter > 1 || throw(ArgumentError("maxiter must be greater than 1."))
         tol > 0 || throw(ArgumentError("tol must be positive."))
-        lambda >= 0 || throw(ArgumentError("lambda must be non-negative."))
-
-        new{T}(obj, maxiter, verbose, tol, lambda)
+        lambda_w >= 0 || throw(ArgumentError("lambda_w must be non-negative."))
+        lambda_h >= 0 || throw(ArgumentError("lambda_h must be non-negative."))
+        if !isnothing(lambda) && lambda >= 0
```
Comment (ararslan) [inline src/multupd.jl:30]:
```suggestion
        if lambda !== nothing && lambda >= 0
```
`isnothing` was added in Julia 1.1 but this package still supports earlier versions. This is why CI is failing.

File: src/multupd.jl (hunk 4 above)
Comment (mschauer) [reply src/multupd.jl:20]: Yes, right. We can name `delta` back to `lambda`, and make something like 
`lambda::Real=sqrt(eps(T)), lambda_w::Real=lambda, lambda_h::Real=lambda` the signature.

Comment (tsano430) [reply src/multupd.jl:20]: @mschauer The original `lambda` (default value 1.0e-9) is probably used only to avoid division by zero. After `delta` is named back to the original `lambda`, setting the original `lambda` to a non-small value may adversely affect the accuracy of the approximation X ≈ WH. Therefore, I don't think `delta` should be user-specified.

Comment (tsano430) [reply src/multupd.jl:20]: The closer `delta` (the original `lambda`) is to zero, the more ideal the algorithm is. 

cf. https://gist.github.com/tsano430/dc005fd59f89662d777774de84cd5d0d

File: src/multupd.jl (hunk 5 above)
Comment (tsano430) [reply src/multupd.jl]: Thank you. I understood it.

**Review by ararslan — APPROVED** [review]

File: src/multupd.jl (hunk 4 above)
Comment (tsano430) [reply src/multupd.jl:20]: Thank you for your reaction with thumbs up emoji.

**Comment by tsano430** [comment]
@ararslan Thank you for your approval!
